release.sh

```diff
@@ -0,0 +1,164 @@
+# derive-macros/ dir) are released.
+
+# Exit on error, undefined variables, and pipe failures
+set -euo pipefail
+
+# print commands before executing them for debugging
+# set -x
+
+RED='\033[0;31m'
+GREEN='\033[0;32m'
+YELLOW='\033[1;33m'
+BLUE='\033[0;34m'
+NC='\033[0m' # no color
+
+log_info() { echo -e "${BLUE}[INFO]${NC} $1"; }
+log_success() { echo -e "${GREEN}[SUCCESS]${NC} $1"; }
+log_warning() { echo -e "${YELLOW}[WARNING]${NC} $1"; }
+log_error() { echo -e "${RED}[ERROR]${NC} $1"; exit 1; }
+
+check_requirements() {
+    log_info "Checking required tools..."
+
+    command -v cargo >/dev/null 2>&1 || log_error "cargo is required but not installed"
+    command -v git >/dev/null 2>&1 || log_error "git is required but not installed"
+    command -v cargo-release >/dev/null 2>&1 || log_error "cargo-release is required but not installed. Install with: cargo install cargo-release"
+    command -v git-cliff >/dev/null 2>&1 || log_error "git-cliff is required but not installed. Install with: cargo install git-cliff"
+    command -v jq >/dev/null 2>&1 || log_error "jq is required but not installed."
+
+    log_success "All required tools are available"
+}
+
+is_main_branch() {
+    local current_branch
+    current_branch=$(git rev-parse --abbrev-ref HEAD)
+    [[ "$current_branch" == "main" ]]
+}
+
+is_working_tree_clean() {
+    git diff --quiet && git diff --cached --quiet
+}
+
+# check if the version is already published on crates.io
+is_version_published() {
+    local crate_name="$1"
+    local version
+    version=get_current_version "$crate_name"
+
+    if [[ -z "$version" ]]; then
+        log_error "Could not find crate '$crate_name' in workspace"
+    fi
+
+    if cargo search "$crate_name" | grep -q "^$crate_name = \"$version\""; then
+        return 0
+    else
+        return 1
+    fi
+}
+
+# get current version from Cargo.toml
+get_current_version() {
+    local crate_name="$1"
+    cargo metadata --no-deps --format-version 1 | \
+        jq -r --arg name "$crate_name" '.packages[] | select(.name == $name) | .version'
+}
+
+# Prompt user for confirmation
+confirm() {
+    local prompt="$1"
+    local response
+
+    echo -e -n "${YELLOW}${prompt} [y/N]${NC} "
+    read -r response
+
+    [[ "$response" =~ ^[Yy] ]]
+}
+
+# handle release branch workflow (CHANGELOG updates, README updates, PR to main)
+handle_release_branch() {
+    local version="$1"
+
+    log_info "Starting release preparation for version $version..."
+
+    # Update CHANGELOG and README
+    log_info "Updating CHANGELOG.md and README.md..."
+    if ! cargo release --workspace "$version" --no-publish --no-push --no-tag --execute; then
+        log_error "Failed to update CHANGELOG and README"
+    fi
+
+    if confirm "Print diff of CHANGELOG/README changes?"; then
+        git diff --stat HEAD^
+        git diff HEAD^
+    fi
+
+    if confirm "Would you like to open a PR with these changes?"; then
```

Comment:
Probably should have two things:
1. Do you want to push these changes
2. Do you want to make a PR

I would want to push, but I don't use `gh`, so I wouldn't want the script to make a PR